server/reviewControllers.js

```diff
@@ -0,0 +1,192 @@
+        const response = {
+          product_id,
+          page,
+          count,
+          result: result.rows.length ? result.rows : [],
+        };
+        res.status(200).json(response);
+      } catch (err) {
+        res.status(500).send('Internal Server Error');
+      }
+    }
+  },
+
+  getMeta: async (req, res) => {
+    if (!req.query.product_id) {
+      res.status(422).send('Invalid product_id');
+    } else {
+      const { product_id } = req.query;
+      const queryText = `
+        SELECT
+          ${product_id} AS product_id,
+          (SELECT json_object_agg(rating::TEXT, ratingCount)
+            FROM
+              (SELECT rating, COUNT(*) AS ratingCount
+                FROM reviews
+                WHERE product_id = ${product_id}
+                GROUP BY rating
+              ) AS rc
+          ) AS ratings,
+          (SELECT json_object_agg(recommend::TEXT, recCount)
+            FROM
+              (SELECT recommend, COUNT (*) AS recCount
+                FROM reviews
+                WHERE product_id = ${product_id}
+                GROUP BY recommend
+              ) AS rec
+          ) AS recommended,
+          (SELECT json_object_agg(name, json_build_object('id', characteristic_id, 'value', avg))
+            FROM
+            (SELECT name, characteristic_id, AVG(value) AS avg
+              FROM characteristics
+              INNER JOIN characteristic_reviews
+              ON characteristics.id = characteristic_reviews.characteristic_id
+              WHERE product_id = ${product_id}
+              GROUP BY name, characteristic_id
+            ) AS char
+          ) AS characteristics
+      `;
+
+      try {
+        const result = await pool.query(queryText);
+        res.status(200).json(result.rows[0]);
+      } catch (err) {
+        res.status(500).send('Internal Server Error');
+      }
+    }
+  },
+
+  addReview: async (req, res) => {
+    if (!req.body.product_id) {
+      res.status(422).send('Invalid product_id');
+    } else {
+      const chars = req.body.characteristics;
+      const queryText = `
+          INSERT INTO reviews(product_id, rating, summary, body, recommend, reviewer_name, reviewer_email)
+          VALUES ($1, $2, $3, $4, $5, $6, $7)
+          RETURNING id;
+      `;
+      const charReviewQueryText = `
+        INSERT INTO characteristic_reviews(characteristic_id, review_id, value)
+        VALUES ($1, $2, $3)
+      `;
+      const photoQueryText = `
+        INSERT INTO photos(review_id, url)
+        VALUES ($1, $2)
+      `;
+
+      try {
+        const result = await pool.query(queryText, [
+          req.body.product_id,
+          req.body.rating,
+          req.body.summary,
+          req.body.body,
+          req.body.recommend,
+          req.body.name,
+          req.body.email,
+        ]);
+        const reviewID = await result.rows[0].id;
+
+        if (Object.keys(chars).length) {
+          Object.entries(chars).forEach(async ([charId, charVal]) => {
```

Comment:
careful using async / await inside of a forEach. the outside context will not wait for the asynchronous code inside the callback to finish before moving to the next line so you will likely be responding with a status of 201 before the promises complete. 
instead use a Promise.all or a regular for loop. 
Or best case scenario for this piece of code would be to construct a JOIN query to get this data at once instead of making extra queries for each characteristic.

Reply:
I appreciate the feedback Burhan! I'll merge the request for now so I can work on my tests. I will definitely try to make changes to this method but if I get stuck, it'd be great if we could go over it tomorrow